numpy/random/mtrand/mvhypergeom.c

```diff
@@ -0,0 +1,246 @@
+        if (j == num_colors - 1) {
+            sample[i + j] = num_to_sample;
+        }
+
+        if (more_than_half) {
+            for (j = 0; j < num_colors; ++j) {
+                sample[i + j] = colors[j] - sample[i + j];
+            }
+        }
+
+        i += num_colors;
+    }
+    return 0;
+}
+
+
+/*
+    Draw variates from the multivariate hypergeometric distribution--
+    the "count" algorithm.
+
+    Parameters
+    ----------
+    rk_state *state
+        Pointer to a randomkit `rk_state` instance.
+    size_t num_colors
+        The length of the `colors` array.
+    long *colors
+        The array of colors (i.e. the number of each type in the collection
+        from which the random variate is drawn).
+    long nsample
+        The number of objects drawn without replacement for each variate.
+        `nsample` must not exceed sum(colors).  This condition is not checked;
+        it is assumed that the caller has already validated the value.
+    size_t num_sample
+        The number of random variates to generate.  (This is not the size of
+        the array `sample`, because each variate requires `num_colors`
+        values.)
+    long *sample
+        The array that will hold the result.  The length of this array must
+        be the product of `num_colors` and `num_sample`.
+        The array is not initialized; it is expected that array has been
+        initialized with zeros when the function is called.
+
+    Notes
+    -----
+    The "count" algorithm for drawing one variate is roughly equivalent to the
+    following numpy code:
+
+        choices = np.repeat(np.arange(len(colors)), colors)
+        selection = np.random.choice(choices, nsample, replace=False)
+        sample = np.bincount(selection, minlength=len(colors))
+
+    This function uses a temporary array with length sum(colors).
+*/
+
+int multivar_hypergeom_count(rk_state *state,
+                             size_t num_colors, long *colors,
+                             long nsample,
+                             size_t num_sample, long *sample)
+{
+    size_t i, j, k;
+    long total;
+    long *choices;
+    int more_than_half;
+
+    if (num_colors < 1) {
+        return -2;
+    }
+
+    /* total = sum(colors) */
+    total = 0;
+    for (i = 0; i < num_colors; ++i) {
+        total += colors[i];
+    }
+
+    choices = (long *) malloc(total * sizeof(long));
+    if (choices == NULL) {
+        return -1;
+    }
+    /*
+     *  If colors contains, for example, [3 2 5], then choices
+     *  will contain [0 0 0 1 1 2 2 2 2 2].
+     */
+    k = 0;
+    for (i = 0; i < num_colors; ++i) {
+        for (j = 0; j < (size_t) colors[i]; ++j) {
+            choices[k] = i;
+            ++k;
+        }
+    }
+
+    more_than_half = nsample > (total / 2);
+    if (more_than_half) {
+        nsample = total - nsample;
+    }
+
+    i = 0;
+    while (i < num_sample) {
+        /*
+         *  Fisher-Yates shuffle, but only loop through the first
```

Comment:
Is Fisher-Yates implemented anywhere else? ISTR that there was some talk of using another algorithm for shuffle.

Reply:
This is basically a random choice without replacement.  Floyd's algorithm can be used (it is mentioned in some older numpy issues and pull requests).  In fact, that is one of the experiments I recently tried: Floyd's algorithm, using khash to implement the set of selected numbers.  However, my implementation didn't provide any clear advantage over the current two implementations, so I abandoned that approach.